File: mobile/ios/FindlyKit/Sources/FindlyKit/LocationSensing/FixCaptureCoordinator.swift
import Foundation

/// specs/009-device-runtime.md §1 (pipeline) + §1.2 (suppression) — the capture-and-queue pipeline,
/// mirroring Android's `FixCaptureCoordinator` split exactly: a thin, CoreLocation-touching
/// `LocationProviding` implementation on one side, and this pure, fake-testable `actor` on the
/// other, which owns:
///
/// - **Suppression** (§1.2): a capture is skipped (not queued, never an error) when tracking is
///   paused, location permission is absent, or an identical-position fix was captured < 60 s ago.
/// - **Ordering**: paused/permission-absent are checked **before** ever invoking the provider (no
///   GPS burn at all); pause is re-checked again **after** capture too, so a pause landing
///   mid-capture drops the result instead of queuing it (specs/009 §4: "in-flight captures dropped,
///   not queued"); the debounce check can only run after a fix comes back (hint or real capture).
///
/// This is **the** seam every I10-owned trigger (the periodic BG-task/significant-location-change
/// path, a future manual-refresh UI call) goes through. `source: .locate` is deliberately **not**
/// routed through here (mirrors Android's `LocateRequestPushHandler` bypassing
/// `FixCaptureCoordinator` and calling `LocationCapturer`/here, `LocationProviding.requestSingleFix`
/// directly) — a `LOCATE_REQUEST` MUST still be fulfilled while paused (specs/009 §5.1), which
/// this class's pause gate would incorrectly block. `hint` is the seam I11's geofence-transition
/// handling calls for its own `source: "geofence"` fix (§6.3: "MAY reuse the transition's own
/// coordinates") and the one `SystemLocationProvider`'s own significant-location-change delegate
/// callback uses (see `LocationProviding.swift`'s doc) — when supplied, `LocationProviding` is
/// never invoked at all, so a hint burns no extra GPS.
public actor FixCaptureCoordinator {
    private let provider: LocationProviding
    private let queue: FixQueue
    private let isPaused: () -> Bool
    private let isPermissionGranted: () -> Bool
    private let now: () -> Date

    /// specs/009 §3.4 (I50 fix 6, Minor) — the same `× 0.8` elapsed-time gate
    /// `LocationSyncRunner.runOnce()`'s `BGAppRefreshTask`/foreground trigger already applies
    /// before calling this method for `source: .periodic`. Optional/`nil`-defaulted: the
    /// significant-location-change/visit hint paths (`SystemLocationProvider`'s
    /// `didUpdateLocations`/`didVisit`, both `LocationProviding.swift`) used to call
    /// `captureAndQueue(source: .periodic, hint:)` directly, bypassing this gate entirely and
    /// leaving `lastQueuedFixAtStore` stale — queuing every hint regardless of interval and making
    /// the very next background-refresh trigger capture again immediately. Applying the gate HERE,
    /// inside the one seam every `.periodic` caller goes through (this class's own doc), fixes both
    /// hint paths for free without either needing to know about `SyncTriggerPolicy` itself. `nil`
    /// (the default) means "no gate" — every existing caller/test that doesn't configure this,
    /// including every non-`.periodic` source, is completely unaffected.
    private let currentSyncIntervalMinutes: (() -> Int)?
    private let lastQueuedFixAtStore: LastQueuedFixAtStoring?

    private var lastCaptured: (lat: Double, lon: Double, at: Date)?

    public init(
        provider: LocationProviding,
        queue: FixQueue,
        isPaused: @escaping () -> Bool,
        isPermissionGranted: @escaping () -> Bool,
        currentSyncIntervalMinutes: (() -> Int)? = nil,
        lastQueuedFixAtStore: LastQueuedFixAtStoring? = nil,
        now: @escaping () -> Date = Date.init
    ) {
        self.provider = provider
        self.queue = queue
        self.isPaused = isPaused
        self.isPermissionGranted = isPermissionGranted
        self.currentSyncIntervalMinutes = currentSyncIntervalMinutes
        self.lastQueuedFixAtStore = lastQueuedFixAtStore
        self.now = now
    }

    /// Captures one fix for `source` and enqueues it, unless suppressed. Returns the
    /// captured/reused fix even when it was ultimately debounce-suppressed post-capture is
    /// intentionally NOT the behavior here (unlike Android's `CapturedFix`-returning variant) —
    /// `nil` always means "nothing was queued", full stop; iOS has no caller today that needs the
    /// position separately from the enqueue outcome (Android's geofence-event handler does, but
    /// that's I11's concern and can widen this return value then if needed).
    @discardableResult
    public func captureAndQueue(source: FixSource, hint: LocationFix? = nil) async -> LocationFix? {
        if isPaused() { return nil }
        if !isPermissionGranted() { return nil }

        // I50 fix 6 (Minor) — the iOS opportunistic-trigger elapsed-time rule (specs/009 §3.4)
        // applies to every `.periodic` caller, hint-triggered ones included; see this gate's own
        // doc for why it lives here rather than only inside `LocationSyncRunner`.
        if source == .periodic, let currentSyncIntervalMinutes, let lastQueuedFixAtStore {
            let shouldCapture = SyncTriggerPolicy.shouldCapture(
                syncIntervalMinutes: currentSyncIntervalMinutes(),
                lastQueuedFixAt: lastQueuedFixAtStore.lastQueuedFixAt(),
                now: now()
            )
            guard shouldCapture else { return nil }
        }

        let fix: LocationFix
        if let hint {
            fix = hint
        } else {
            guard let captured = try? await provider.requestSingleFix(source: source) else { return nil }
            fix = captured
        }

        // Pause arriving mid-capture: drop, don't queue (specs/009 §4).
        if isPaused() { return nil }

        if isDuplicateWithinDebounce(fix) { return nil }
        lastCaptured = (fix.lat, fix.lon, now())

        await queue.enqueue(fix)
        // I50 fix 6 — record on every successful `.periodic` enqueue (hint-triggered included), so
        // the very next background-refresh trigger correctly re-evaluates the same 0.8 threshold
        // instead of finding a stale timestamp and capturing again immediately.
        if source == .periodic {
            lastQueuedFixAtStore?.recordQueuedFixAt(now())
        }
        return fix
    }

    /// specs/009 §1.2: "an identical-position fix was captured < 60 s ago (debounce against
    /// duplicate platform callbacks)."
    private func isDuplicateWithinDebounce(_ fix: LocationFix) -> Bool {
        guard let lastCaptured else { return false }
        let samePosition = lastCaptured.lat == fix.lat && lastCaptured.lon == fix.lon
        let withinDebounceWindow = now().timeIntervalSince(lastCaptured.at) < 60
        return samePosition && withinDebounceWindow
    }
}

File: mobile/android/app/src/main/java/com/findly/android/location/FixCaptureCoordinator.kt
package com.findly.android.location

import com.findly.android.queue.FixQueueStore
import com.findly.android.queue.FixSource
import com.findly.android.queue.QueuedFix
import java.time.Duration
import java.time.Instant
import java.util.UUID

/** Whether tracking is currently paused (`trackingEnabled: false`, specs/009-device-runtime.md
 * §4) — backed by whatever holds the last-applied [com.findly.android.network.DeviceSettingsSnapshot]. */
fun interface TrackingPauseState {
    suspend fun isPaused(): Boolean
}

/** Whether this device currently holds the location permission a capture needs (specs/009 §7). */
fun interface LocationPermissionState {
    suspend fun isGranted(): Boolean
}

/**
 * The capture-and-queue pipeline (specs/009-device-runtime.md §1) — ties A9's [LocationCapturer]
 * seam + [FixQueueStore] together with the §1.2 suppression rules, and owns the
 * [FixSource] → ([LocationAccuracyTier], timeout) mapping A9's seam deliberately left to its
 * callers (see [LocationCapturer]'s doc for why `BALANCED` alone can't tell `periodic` and
 * `geofence` apart). This is **the** seam every A10-owned trigger calls through: the periodic
 * worker/foreground service (`source: "periodic"`) and a manual refresh (`source: "manual"`).
 * `source: "locate"` is deliberately **not** routed through here — A9's
 * [com.findly.android.pushmessages.LocateRequestPushHandler] calls [LocationCapturer] directly,
 * bypassing suppression entirely, because a `LOCATE_REQUEST` must still be fulfilled while paused
 * (§5.1) — this class's pause/permission gate would be actively wrong for that case.
 *
 * `hint` is the seam **A11**'s geofence-transition handling calls for its own `source: "geofence"`
 * fix (§6.3: "additionally capture one fix with `source: geofence`") — when supplied, the
 * transition's own coordinates are used directly and [LocationCapturer] is never invoked at all
 * (§1.1: "MAY reuse the transition's own coordinates").
 *
 * Suppression order matters (§1.2): paused / permission-absent are checked **before** ever
 * invoking the capturer (no GPS burn at all); the identical-position debounce can only be checked
 * **after** a fix comes back, and pause is re-checked after capture too, so a pause that lands
 * mid-capture drops the result instead of queuing it ("in-flight captures dropped, not queued",
 * 009 §4). Every skip path is silent — never an exception, never surfaced to the user (§1.2).
 */
class FixCaptureCoordinator(
    private val capturer: LocationCapturer,
    private val queueStore: FixQueueStore,
    private val pauseState: TrackingPauseState,
    private val permissionState: LocationPermissionState,
    private val fixIdGenerator: () -> String = { UUID.randomUUID().toString() },
    private val now: () -> Instant = Instant::now,
) {
    private var lastCaptured: Pair<CapturedFix, Instant>? = null

    /**
     * Captures one fix for [source] and enqueues it, unless suppressed. [hint] short-circuits an
     * actual GPS request (see class doc). Returns the captured/reused fix even when it was
     * ultimately debounce-suppressed post-capture, so a caller like a geofence-event handler can
     * still use the position for its own purposes — `null` only ever means "nothing was queued".
     */
    suspend fun captureAndQueue(source: FixSource, hint: CapturedFix? = null): CapturedFix? {
        if (pauseState.isPaused()) return null
        if (!permissionState.isGranted()) return null

        val fix = hint ?: capturer.captureFix(accuracyFor(source), timeoutMillisFor(source)) ?: return null

        // Pause arriving mid-capture: drop, don't queue (009 §4).
        if (pauseState.isPaused()) return null

        if (isDuplicateWithinDebounce(fix)) return null
        lastCaptured = fix to now()

        queueStore.enqueue(
            QueuedFix(
                fixId = fixIdGenerator(),
                recordedAt = fix.recordedAt,
                lat = fix.lat,
                lon = fix.lon,
                accuracyM = fix.accuracyM,
                altitudeM = fix.altitudeM,
                speedMps = fix.speedMps,
                bearingDeg = fix.bearingDeg,
                batteryPct = fix.batteryPct,
                source = source,
            ),
        )
        return fix
    }

    /** §1.2: "an identical-position fix was captured < 60 s ago (debounce against duplicate
     * platform callbacks)". */
    private fun isDuplicateWithinDebounce(fix: CapturedFix): Boolean {
        val (previousFix, previousAt) = lastCaptured ?: return false
        val samePosition = previousFix.lat == fix.lat && previousFix.lon == fix.lon
        val withinDebounceWindow = Duration.between(previousAt, now()) < DEBOUNCE_WINDOW
        return samePosition && withinDebounceWindow
    }

    private companion object {
        val DEBOUNCE_WINDOW: Duration = Duration.ofSeconds(60)
        const val DEFAULT_TIMEOUT_MILLIS = 30_000L
        const val GEOFENCE_TIMEOUT_MILLIS = 15_000L

        /** §1.1's "Accuracy request" column for the sources that reach this class. */
        fun accuracyFor(source: FixSource): LocationAccuracyTier = when (source) {
            FixSource.Manual -> LocationAccuracyTier.HIGH
            FixSource.Periodic, FixSource.Geofence, FixSource.Locate -> LocationAccuracyTier.BALANCED
        }

        /** §1.1's "Timeout" column — the only reason this class doesn't just rely on
         * [LocationCapturer.captureFix]'s own 30 s default is `geofence`'s shorter 15 s. */
        fun timeoutMillisFor(source: FixSource): Long = when (source) {
            FixSource.Geofence -> GEOFENCE_TIMEOUT_MILLIS
            else -> DEFAULT_TIMEOUT_MILLIS
        }
    }
}
